File: src-tauri/src/commands/share.rs
//! 组网（TokenTap Share）相关的 Tauri 命令

use crate::share::keystore::KeyStorage;
use crate::share::types::{
    CreateNetworkResult, RequestJoinResult, ShareNetworkStatus, ShareQuotaConfig,
};
use crate::store::AppState;

/// 创建组网网络，返回 share id 与 share key（key 仅展示一次）
#[tauri::command]
pub async fn share_create_network(
    state: tauri::State<'_, AppState>,
    node_name: Option<String>,
) -> Result<CreateNetworkResult, String> {
    let name = node_name
        .filter(|n| !n.trim().is_empty())
        .unwrap_or_else(|| {
            // 默认节点名：主机名，退化为“未命名节点”
            hostname::get()
                .ok()
                .and_then(|h| h.into_string().ok())
                .filter(|s| !s.is_empty())
                .unwrap_or_else(|| "未命名节点".to_string())
        });
    state.share_manager.create_network(name).await
}

/// 发起加入（凭 share id），返回 6 位短码等待出借方审批
#[tauri::command]
pub async fn share_request_join(
    state: tauri::State<'_, AppState>,
    share_id: String,
) -> Result<RequestJoinResult, String> {
    state.share_manager.request_join(share_id).await
}

/// 取消等待中的加入申请
#[tauri::command]
pub async fn share_cancel_join(state: tauri::State<'_, AppState>) -> Result<(), String> {
    state.share_manager.cancel_join().await
}

/// 出借方批准加入申请
#[tauri::command]
pub async fn share_approve_join(
    state: tauri::State<'_, AppState>,
    peer_id: String,
) -> Result<(), String> {
    state.share_manager.approve_join(&peer_id).await
}

/// 出借方拒绝加入申请
#[tauri::command]
pub async fn share_reject_join(
    state: tauri::State<'_, AppState>,
    peer_id: String,
    reason: Option<String>,
) -> Result<(), String> {
    state.share_manager.reject_join(&peer_id, reason).await
}

/// 退出/解散网络
#[tauri::command]
pub async fn share_leave_network(state: tauri::State<'_, AppState>) -> Result<(), String> {
    state.share_manager.leave_network().await
}

/// 获取组网完整状态
#[tauri::command]
pub async fn share_get_status(
    state: tauri::State<'_, AppState>,
) -> Result<ShareNetworkStatus, String> {
    state.share_manager.get_status().await
}

/// 设置出借白名单（"app:provider_id" 列表）
#[tauri::command]
pub async fn share_set_shared_providers(
    state: tauri::State<'_, AppState>,
    provider_ids: Vec<String>,
) -> Result<(), String> {
    state.share_manager.set_shared_providers(provider_ids).await
}

/// 设置出借限额
#[tauri::command]
pub async fn share_set_quota(
    state: tauri::State<'_, AppState>,
    config: ShareQuotaConfig,
) -> Result<(), String> {
    state.share_manager.set_quota(config).await
}

/// 拉黑节点
#[tauri::command]
pub async fn share_block_peer(
    state: tauri::State<'_, AppState>,
    peer_id: String,
    reason: Option<String>,
) -> Result<(), String> {
    state.share_manager.block_peer(&peer_id, reason).await
}

/// 解除拉黑
#[tauri::command]
pub async fn share_unblock_peer(
    state: tauri::State<'_, AppState>,
    peer_id: String,
) -> Result<(), String> {
    state.share_manager.unblock_peer(&peer_id).await
}

/// 重新生成 share key（旧 key 立即失效）
#[tauri::command]
pub async fn share_regenerate_key(state: tauri::State<'_, AppState>) -> Result<String, String> {
    state.share_manager.regenerate_key().await
}

/// 设置消费侧路由偏好（local_only/local_first/network_first/network_only）
#[tauri::command]
pub async fn share_set_route_preference(
    state: tauri::State<'_, AppState>,
    preference: String,
) -> Result<(), String> {
    state.share_manager.set_route_preference(preference).await
}

/// 设置 relay 地址覆盖（None = 恢复官方默认）
#[tauri::command]
pub async fn share_set_relay_addr(
    state: tauri::State<'_, AppState>,
    addr: Option<String>,
) -> Result<(), String> {
    state.share_manager.set_relay_addr(addr).await
}

/// 设置节点显示名
#[tauri::command]
pub async fn share_set_node_name(
    state: tauri::State<'_, AppState>,
    name: String,
) -> Result<(), String> {
    state.share_manager.set_node_name(name).await
}

/// share key 的存储位置（keyring/file，file 时 UI 提示风险）
#[tauri::command]
pub async fn share_key_storage(state: tauri::State<'_, AppState>) -> Result<KeyStorage, String> {
    Ok(state.share_manager.key_storage().await)
}

File: src/lib/api/share.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  CreateNetworkResult,
  RequestJoinResult,
  ShareKeyStorage,
  ShareQuotaConfig,
  ShareProviderCheckResult,
  ShareStatus,
} from "@/types/share";

/** 组网（TokenTap Share）API */
export const shareApi = {
  // ========== 网络生命周期 ==========

  /** 创建网络（返回 share id 与仅展示一次的 share key） */
  async createNetwork(nodeName?: string): Promise<CreateNetworkResult> {
    return invoke("share_create_network", { nodeName: nodeName ?? null });
  },

  /** 发起加入（返回 6 位短码，等待出借方审批） */
  async requestJoin(shareId: string): Promise<RequestJoinResult> {
    return invoke("share_request_join", { shareId });
  },

  /** 取消等待中的加入申请 */
  async cancelJoin(): Promise<void> {
    return invoke("share_cancel_join");
  },

  /** 退出/解散网络 */
  async leaveNetwork(): Promise<void> {
    return invoke("share_leave_network");
  },

  // ========== 审批（出借方） ==========

  async approveJoin(peerId: string): Promise<void> {
    return invoke("share_approve_join", { peerId });
  },

  async rejectJoin(peerId: string, reason?: string): Promise<void> {
    return invoke("share_reject_join", { peerId, reason: reason ?? null });
  },

  // ========== 状态 ==========

  async getStatus(): Promise<ShareStatus> {
    return invoke("share_get_status");
  },

  /** share key 存储位置（file 时 UI 提示风险） */
  async keyStorage(): Promise<ShareKeyStorage> {
    return invoke("share_key_storage");
  },

  // ========== 出借方控制 ==========

  /** 设置出借白名单（"app:provider_id" 列表） */
  async setSharedProviders(providerIds: string[]): Promise<void> {
    return invoke("share_set_shared_providers", { providerIds });
  },

  /** 设置限额 */
  async setQuota(config: ShareQuotaConfig): Promise<void> {
    return invoke("share_set_quota", { config });
  },

  /** 拉黑节点 */
  async blockPeer(peerId: string, reason?: string): Promise<void> {
    return invoke("share_block_peer", { peerId, reason: reason ?? null });
  },

  /** 解除拉黑 */
  async unblockPeer(peerId: string): Promise<void> {
    return invoke("share_unblock_peer", { peerId });
  },

  /** 重新生成 share key（旧 key 立即失效） */
  async regenerateKey(): Promise<string> {
    return invoke("share_regenerate_key");
  },

  // ========== 配置 ==========

  /** 设置消费侧路由偏好 */
  async setRoutePreference(preference: string): Promise<void> {
    return invoke("share_set_route_preference", { preference });
  },

  async setRouteMode(mode: "provider" | "consumer"): Promise<void> {
    return invoke("share_set_route_mode", { mode });
  },

  /** 设置某个应用实际参与路由的共享 Provider target */
  async setRouteTargets(appType: string, targets: string[]): Promise<void> {
    return invoke("share_set_route_targets", { appType, targets });
  },

  /** 启用共享 Provider，并同步 Agent live 配置与本地代理接管。 */
  async activateProvider(
    appType: string,
    peerId: string,
    providerId: string,
  ): Promise<void> {
    return invoke("share_activate_provider", {
      appType,
      peerId,
      providerId,
    });
  },

  /** 检测远端共享 Provider 连通性 */
  async testProvider(
    appType: string,
    peerId: string,
    providerId: string,
  ): Promise<ShareProviderCheckResult> {
    return invoke("share_test_provider", { appType, peerId, providerId });
  },

  /** 设置 relay 地址（null = 恢复官方默认） */
  async setRelayAddr(addr: string | null): Promise<void> {
    return invoke("share_set_relay_addr", { addr });
  },

  /** 设置节点显示名 */
  async setNodeName(name: string): Promise<void> {
    return invoke("share_set_node_name", { name });
  },
};
